teps/0144-param-enum.md

```diff
@@ -0,0 +1,314 @@
+- Tekton should give early and explicit error message if the Task/Pipeline is failed due to invalid `param` input
+
+## Proposal
+We propose adding an ***optional*** field named `enum` to the [`ParamSpec`](https://github.com/tektoncd/pipeline/blob/main/docs/pipeline-api.md#paramspec) in a [`PipelineSpec`](https://github.com/tektoncd/pipeline/blob/main/docs/pipeline-api.md#tekton.dev/v1.PipelineSpec) or [`TaskSpec`](https://github.com/tektoncd/pipeline/blob/main/docs/pipeline-api.md#tekton.dev/v1.TaskSpec). This field must be an array with at least one element, where each element is unique.
+
+> :warning: The new API field is introduced as an `alpha` feature
+
+> :warning: This `enum` field should be specified if and only if the parameter type is `string`, which is similar to the existing [`properties`](https://github.com/tektoncd/pipeline/blob/main/docs/pipeline-api.md#paramspec) field that should be only specified for `object` type.
+
+### Tasks
+The following is a `Task` example with the `task-version` param bounded to a set of constants: `v1.21`, `v1.20` and `v1.19`: 
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Task
+metadata:
+  name: golang-build
+spec:
+  params:
+  - name: task-version
+    description: golang version to use for builds
+    type: string
+    enum: # this field can be specified iff the type is string
+      - "v1.21"
+      - "v1.20"
+      - "v1.19"
+    default: "v1.21"
+  steps:
+  - name: build
+    image: docker.io/library/golang:$(params.task-version)
+    ...
+```
+
+#### TaskRuns with TaskRef or Remote Resolution
+`Task` users can provide a value for the parameter at the execution time. If the value is in the `enum` list specified by the author, the task will be executed normally. The task will fail the validation with reason `InvalidParamValue` otherwise. 
+
+As an example, the `TaskRun` references the above `Task`. The execution of the `TaskRun` should fail with reason `InvalidParamValue` and a detailed error message since `latest` is NOT in the `enum` list. 
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: TaskRun
+metadata:
+  name: run-golang-build
+spec:
+  params:
+  - name: version
+    value: "latest" # 'latest' is not in the enum list
+  taskRef:
+    name: golang-build
+    ...
+```
+
+#### TaskRuns with Embedded TaskSpec
+`Task` users are allowed to specify `enum` for `TaskRuns` with Embedded `TaskSpec` and Tekton will perform the same validation as described above.
+
+``` yaml
+apiVersion: tekton.dev/v1
+kind: TaskRun
+metadata:
+  generateName: golang-build-run
+spec:
+  params:
+    - name: version
+      value: "v1.21"
+  taskSpec:
+    params:
+      - name: version
+        type: string 
+        enum:
+          - "v1.21"
+          - "v1.20"
+          - "v1.19"
+    ...
+```
+
+While we haven't identified any `enum` use case in this scenario at the time writing this TEP, this is treated as a valid `TaskRun` for `TaskSpec` API compatibility reason.
```

Comment:
Does this example imply that Task Enum take precedence over TaskRun embedded spec Enum?

Reply:
Hi @JeromeJu, In this scenario, the Task user is the Task author, so this is like the task user specifies an enum for him/herself. However, if the enum is specified, the user provided value in TaskRun (v1.21) must be in the enum list (v1.21, v1.20, v1.19) to pass the validation, which is the same validation logic as described above.

Reply:
This does not provide much value and might create confusion

Reply:
Thanks, @pritidesai for the input!

Yeah, I agree that this does not provide real value from usability point of view since the Task author is the Task user. 👍 
And I think the same thing also applies to using [ParamSpec.Default](https://github.com/tektoncd/pipeline/blob/5d403c6f04ea9ada76f8c4392efb3b73ca332d8b/pkg/apis/pipeline/v1/param_types.go#L55) in embedded `TaskRun`, where there is no clear use case to specify a default value for myself.

The question is that should we fail the validation and treat such `TaskRuns` as invalid?

From API **compatibility** perspective,  we want to keep [taskrun.TaskSpec](https://github.com/tektoncd/pipeline/blob/5d403c6f04ea9ada76f8c4392efb3b73ca332d8b/pkg/apis/pipeline/v1/taskrun_types.go#L46C2-L47C48) and [Task.Spec](https://github.com/tektoncd/pipeline/blob/5d403c6f04ea9ada76f8c4392efb3b73ca332d8b/pkg/apis/pipeline/v1/task_types.go#L44C1-L46C29) compatible as they references the same [TaskSpec](https://github.com/tektoncd/pipeline/blob/5d403c6f04ea9ada76f8c4392efb3b73ca332d8b/pkg/apis/pipeline/v1/task_types.go#L78) struct, as we apply the same validation logic to `Taskrun.TaskSpec` and `Task.Spec`. So I'd say we can still treat it as a valid syntax and don't break this compatibility even though it is a bit redundant. 

From usability point perspective, I don't think it will really confuse users as this is an **optional** field, and I don't expect the author to specify such value at all for embedded TaskRuns.

WDYT?

Reply:
This discussion is also related to embedded PipelineRun and Pipeline with embedded Tasks

Reply:
API WG update:

Allowing enum and applying validation in Taskrun with inline TaskSpec is not "harmful" since this is an optional field. We will implement as designed to favor compatibility.

(Will also add more explanation to the TEP)

Reply:
I've added a bit more information to address user confusion in the TEP, PTAL